fix(compiler-cli): do not type check native controls with ControlValueAccessor

crisbeto opened the pull request:
Currently when we detect a `field` binding on a native element, we treat it as a built-in native control. This might not be the case if it's a pre-existing `ControlValueAccessor` relying on the CVA interop.

These changes try to detect any CVA-like directive on the element and disable the additional type checking if there are any.

Fixes #65468.

crisbeto committed:
refactor(compiler-cli): track public methods during analysis

Updates the directive analysis to track the public methods of the class.

mmalerba commented on packages/compiler-cli/src/ngtsc/typecheck/src/ops/scope.ts:
```diff
@@ -776,7 +777,9 @@ export class Scope {
       node instanceof TmplAstElement &&
       (node.name === 'input' || node.name === 'select' || node.name === 'textarea') &&
-      !allDirectiveMatches.some(getCustomFieldDirectiveType)
+      allDirectiveMatches.every((meta) => {
```
nit: This `if` logic is getting pretty long, would it make sense to extract the new condition into a function called `hasCVA` or something

crisbeto committed:
fix(compiler-cli): do not type check native controls with ControlValueAccessor

Currently when we detect a `field` binding on a native element, we treat it as a built-in native control. This might not be the case if it's a pre-existing `ControlValueAccessor` relying on the CVA interop.

These changes try to detect any CVA-like directive on the element and disable the additional type checking if there are any.

Fixes #65468.

leonsenft reviewed (approved):
Are CVAs always directives? I know they're generally provided via the `NG_VALUE_ACCESSOR` token, so is there an edge case where the provided accessor isn't actually a directive, and could we catch that in type checking or no?
